pkg/interceptors/client/logging.go

```diff
@@ -0,0 +1,81 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"go.uber.org/zap"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/status"
+)
+
+// LoggingInterceptor logs errors for unary and stream RPCs.
+type LoggingInterceptor struct {
+	logger *zap.Logger
+}
+
+// NewLoggingInterceptor creates a new instance of LoggingInterceptor.
+func NewLoggingInterceptor(logger *zap.Logger) (*LoggingInterceptor, error) {
+	if logger == nil {
+		return nil, fmt.Errorf("logger is required")
+	}
+
+	return &LoggingInterceptor{
+		logger: logger,
+	}, nil
+}
+
+// Unary intercepts unary RPC calls to log errors.
+func (i *LoggingInterceptor) Unary() grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (interface{}, error) {
+		start := time.Now()
+		resp, err := handler(ctx, req) // Call the actual RPC handler
+		duration := time.Since(start)
+
+		if err != nil {
+			st, _ := status.FromError(err)
```

Comment:
Are we 100% sure the error will always be parse-able by `status.FromError`? I know GRPC converts any error eventually to a GRPC unknown error, but I'm less sure it happens _before_ the interceptors run.

If we got another error type the `st.Code()` call below would panic, since `st` would be nil.

Reply:
Yes. It never returns `nil`
```
func FromError(err error) (s *Status, ok bool) {
	if err == nil {
		return nil, true
	}
	type grpcstatus interface{ GRPCStatus() *Status }
	if gs, ok := err.(grpcstatus); ok {
		grpcStatus := gs.GRPCStatus()
		if grpcStatus == nil {
			// Error has status nil, which maps to codes.OK. There
			// is no sensible behavior for this, so we turn it into
			// an error with codes.Unknown and discard the existing
			// status.
			return New(codes.Unknown, err.Error()), false
		}
		return grpcStatus, true
	}
	var gs grpcstatus
	if errors.As(err, &gs) {
		grpcStatus := gs.GRPCStatus()
		if grpcStatus == nil {
			// Error wraps an error that has status nil, which maps
			// to codes.OK.  There is no sensible behavior for this,
			// so we turn it into an error with codes.Unknown and
			// discard the existing status.
			return New(codes.Unknown, err.Error()), false
		}
		p := grpcStatus.Proto()
		p.Message = err.Error()
		return status.FromProto(p), true
	}
	return New(codes.Unknown, err.Error()), false
}
```

Reply:
Perfect. The `FromError` will always coerce into the right type.